Posts/2021/07/How Can I Be Notified Any Time a Service Goes Down.md

```diff
@@ -0,0 +1,209 @@
+This post uses the latest version of PowerShell 7.
+So if you are don't yet have PowerShell 7, get it for free from [Microsoft.com](http://microsoft.com/).
+
+Also, make sure that PowerShell is running as administrator.
+
+## Finding the Required Class
+
+Before going into details, you need to find the required class to monitor.
+To get a list of all available classes, use the following code.
+
+```powershell-console
+Get-CimClass -Namespace root\cimv2
+```
+
+The returned result represents all the available classes in the namespace.
+To monitor Windows services, the WMI class name is *Win32_Service*.
+
+To Enumerate the *Win32_Services* WMI class and get all the available services using PowerShell run the following code.
+
+```powershell-console
+PS C:\> Get-CimInstance -Namespace root\CIMV2 -ClassName win32_service
+```
+
+The result looks like the following.
+
+```powershell-console
+PS C:\> Get-CimInstance -Namespace root\CIMV2 -ClassName win32_service
+```
+
+```powershell
+PS C:\> Get-CimInstance -Namespace root\CIMV2 -ClassName win32_service
+
+ProcessId Name                                     StartMode State   Status  ExitCode
+--------- ----                                     --------- -----   ------  --------
+3784      AdobeARMservice                          Auto      Running OK      0
+3792      AdobeUpdateService                       Auto      Running OK      0
+3800      AGMService                               Auto      Running OK      0
+3824      AGSService                               Auto      Running OK      0
+0         AJRouter                                 Manual    Stopped OK      1077
+0         ALG                                      Manual    Stopped OK      1077
+0         AppIDSvc                                 Manual    Stopped OK      1077
+6708      Appinfo                                  Manual    Running OK      0
+21444     AppMgmt                                  Manual    Running OK      0
+0         AppReadiness                             Manual    Stopped OK      1077
+0         AppVClient                               Disabled  Stopped OK      1077
+0         AppXSvc                                  Manual    Stopped OK      0
+0         AssignedAccessManagerSvc                 Manual    Stopped OK      1077
+```
+
+For now, you can use PowerShell to find the required class, and enumerate it to get the available services.
+Let's go deeper now and start creating the WMI Event Subscription.
+
+There are three steps you need to follow to create a WMI Event Subscription:
+
+- Create the WMI query
+- Registering the query
+- Reading the events.
+
+## Creating WMI Query
+
+WMI database has several system WMI Classes.
+For instance, the *CIM_InstModification* monitors the targeted class for any changes.
+So, in this case, *Win32_Service*.
+
+The query syntax looks like
+
+```powershell-console
+Select * from <WMI System Class> within <Number of Seconds> where TargetInstance ISA <WMI Class name>

```

Comment:
SELECT * FROM <WMI System Class> WITHIN <Number of Seconds> WHERE TargetInstance ISA <WMI Class name>

Reply:
Is it ok to keep the select statement as I explain it, from my point of view it show the reader where the system class should placed and what is the number that comes after WITHIN...
or you prefer write it as **SELECT * FROM WITHIN WHERE TargetInstance ISA**
Thanks